bridges/modules/proof-root-store/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This pallet provides mechanisms for tracking and updating data related to the state root of an
+//! external chain.
+//!
+//! The `T::Value` type represents state root-related data—for example, a `state_root` or
+//! an entire `HeadData` structure.
+//! The `T::Key` type serves as the identifier in the map where we store `T::Value`.
+//! For example, it could be a `block_hash` or `block_number`.
+//!
+//! Example use cases:
+//! 1. Store a `block_hash` → `state_root` mapping.
+//! 2. Store a `block_number` → `HeadData` mapping.
+//!
+//! Root data is stored in a ring buffer, respecting the `T::RootsToKeep` limit.
+//! When the limit is reached, the oldest entries are removed.
+//!
+//! There are two approaches for storing data:
+//! 1. Send root data between chains using the dedicated extrinsic `fn note_new_roots(...)`.
+//! 2. Implement an adapter (e.g., using the `OnSystemEvent` callback) that triggers
+//!    `pallet_bridge_proof_root_store::Pallet::<T, I>::do_note_new_roots(...)`.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+use alloc::collections::VecDeque;
+
+pub use pallet::*;
+pub mod weights;
+pub use weights::WeightInfo;
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarking {
+	// TODO: FAIL-CI
+}
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	/// Configuration trait for the pallet.
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// The origin allowed to submit head updates.
+		type SubmitOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The key type used to identify stored values of type `T::Value`.
+		type Key: Parameter + MaxEncodedLen;
+
+		/// The type of the root value.
+		type Value: Parameter + MaxEncodedLen;
+
+		/// Maximum number of roots to retain in storage.
+		/// This setting prevents unbounded growth of the on-chain state.
+		#[pallet::constant]
+		type RootsToKeep: Get<u32>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	/// Storage for root-related data, bounded by `RootIndex` and `T::RootsToKeep`.
+	#[pallet::storage]
+	pub type Roots<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Blake2_128Concat, T::Key, T::Value, OptionQuery>;
+
+	/// Storage tracking the insertion order of roots for `T::RootsToKeep` (implemented as a simple
+	/// ring buffer).
```

Comment:
We already use ring buffers in the grandpa and parachains pallets. This would be another use for it. And if I understood correctly we might need something like this for emergency governance as well ?

In the past I investigated if this can be deduplicated, but I didn't know much at that point and I thought it wasn't worth it: https://github.com/paritytech/parity-bridges-common/issues/1599. But now, if we need one more ring buffer I think it would be worth investigating that again.

Maybe we could deduplicate this ring buffers logic. Either by having another storage type for ring buffers or by having some helpers